### DISCUSSION TITLE
Utilizing GPU on MacOS for performance

### DISCUSSION THREAD

**Discussion by Darveesh** [TDAmeritrade/stumpy#307] [Q&A]
In a separate discussion thread [#305](https://github.com/TDAmeritrade/stumpy/discussions/305) I am trying to using ``stumpy.mstump``.   My data set is about ~4K "uneven" timestamped geolocations (lat/lon).  After interpolating the data to 1 second increments, the data set becomes ~800K rows.  Feeding that into ``stumpy.mstump`` is taking a long time however.  I started it about 40 min ago and it's still going.  Running on a new/2019 MacBook Pro 16.  Is there an option to do mstump using onboard GPU to speed up the process perhaps?  Seems I have an AMD Radeon Pro 5300M in case it matters.

**Comment by seanlaw** [comment]
@Darveesh Unfortunately, `mstump` has not been ported to GPUs yet as the algorithm is a bit more complex than the 1-dimensional case and, having said that, our GPU algorithms only support NVIDIA GPUs and not AMD so you're out of luck either way.

> After interpolating the data to 1 second increments, the data set becomes ~800K rows. 

800K rows is quite a lot of data and if I were to estimate/extrapolate, based on this [performance table](https://github.com/TDAmeritrade/stumpy#performance), it would take a few hours to complete and this is dependent on the number of cores and amount of memory available on your system. If we're talking about bus data, I'm not sure how much value there is in analyzing at a 1 second granularity. Instead, I would try downsampling the data to something like 5-10 second intervals (and thereby decreasing your data size). At the end of the day, you are interested in path deviations and you are not likely to notice those deviations even at the second level as deviations are likely to occur at the minutely level. 13K rows (of minutely) data should be possible. A quick test on my modest 2-core MBP shows:

```
import numpy as np
import stumpy
import time

T = np.random.rand(30000).reshape((2, 15000))  # Roughly downsampling your data to every minute
start = time.time()
mp, idx = stumpy.mstump(T, m=50)
print(time.time() - start)
# 12.7 seconds

T = np.random.rand(160000).reshape((2, 80000))  # Roughly downsampling your data to every 10 seconds
start = time.time()
mp, idx = stumpy.mstump(T, m=50)
print(time.time() - start)
# 236.9 seconds
```

I hope that helps

**Reply by Darveesh** [reply]
Yes that helps!  I agree, the second by second data set is not relevant or applicable in this case.  I'll resample the data into 1 min "bins" perhaps taking the last value (in case my data set has multiple rows in the same minute).  Thank you.

**Reply by seanlaw** [reply]
Please report back if you find anything interesting!

**Reply by Darveesh** [reply]
Ps: While I was awaiting your input, I came across [PlaidML](https://github.com/plaidml/plaidml)  A project that allows keras and tensorflow to utilize non-Nvidia GPUs.  Just passing that along in case you can utilize it for broader support in future.

**Reply by seanlaw** [reply]
In case you missed it, STUMPY also offers a distributed version of `mstump` called `mstumped` (notice the appnded `ed`) that leverages a Dask distributed (CPU) cluster. If you happen to have a Dask cluster available then you can easily scale up your computation and distribute it across a cluster of servers. We've tested this up to 32 servers (256 CPUs) without breaking a sweat though you will have to refer to the Dask distributed documentation for installation support.
